sdk/main/src/MnemonicBIP39.cc

```diff
@@ -0,0 +1,192 @@
+  // total array size is 17 or 33, for 12 and 24 word mnemonics respectively.
+  //
+  // Algorithm start:
+  // (1) The 8 bits of the first byte in the entropyAndChecksum array are `or`ed into 32 bit scratch
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -> XXXXXXXXXXXXXXXXXXXXXXXX________
+  //                         ^^^^^^^^
+  //                           byte
+  //
+  // Each word index being produced is composed of 11 bits. Therefore, the first byte alone isn't enough to produce the
+  // first index
+  //
+  // (2) Scratch is shifted left by 8 bits to make room for the next byte
+  // XXXXXXXXXXXXXXXXXXXXXXXX________ -> XXXXXXXXXXXXXXXX________XXXXXXXX
+  //                         ^^^^^^^^
+  //                    must be preserved
+  //
+  // (1) The next byte is `or`ed in
+  // XXXXXXXXXXXXXXXX________XXXXXXXX -> XXXXXXXXXXXXXXXX________________
+  //                         ^^^^^^^^
+  //                         next byte
+  //
+  // (3) Once at least 11 bits have been set, it is time to peel off the first index
+  // XXXXXXXXXXXXXXXX_____(___________) -> XXXXXXXXXXXXXXXXXXXXXXXXXXX_____
+  //                       ^^^^^^^^^^^
+  //         convert to int and add to index list
+  //
+  // Repeat until all bytes have been handled
+
+  std::vector<uint16_t> wordIndicesOut = {};
+
+  unsigned int scratch = 0;
+  unsigned int offset = 0;
+  for (unsigned char byte : entropyAndChecksum)
+  {
+    scratch <<= 8;   // (2)
+    scratch |= byte; // (1)
+
+    offset += 8;
+
+    if (offset >= 11) // (4)
+    {
+      // zero out all bits except the 11 we care about
+      wordIndicesOut.push_back((scratch >> (offset - 11)) & 0x7FF);
+      offset -= 11;
+    }
+  }
+
+  return wordIndicesOut;
+}
+
+std::vector<unsigned char> MnemonicBIP39::toSeed(const std::string& passphrase) const
+{
+  EVP_MD_CTX* messageDigestContext = EVP_MD_CTX_new();
+
+  if (!messageDigestContext)
+  {
+    throw std::runtime_error("Digest context construction failed");
+  }
+
+  EVP_MD* messageDigest = EVP_MD_fetch(nullptr, "SHA512", nullptr);
+
+  if (messageDigest == nullptr)
+  {
+    throw std::runtime_error("Digest construction failed");
+  }
+
+  if (EVP_DigestInit(messageDigestContext, messageDigest) <= 0)
+  {
+    throw std::runtime_error("Digest init failed");
+  }
+
+  std::vector<unsigned char> seed(64);
+
+  const std::string mnemonicString = toString();
+  const char* mnemonicStringAddress = mnemonicString.c_str();
+
+  const std::string salt = "mnemonic" + passphrase;
+  const unsigned char* saltAddress = (unsigned char*)salt.c_str();
+
+  if (PKCS5_PBKDF2_HMAC(mnemonicStringAddress,
+                        (int)mnemonicString.length(),
+                        saltAddress,
+                        (int)salt.length(),
+                        2048,
+                        messageDigest,
+                        (int)seed.size(),
+                        &seed.front()) <= 0)
+  {
+    throw std::runtime_error("PKCS5_PBKDF2_HMAC failed");
+  }
```

Comment:
Any if-block that throws in this function should `EVP_MD_CDTX_free(messageDigestContext)` and/or `EVP_MD_free(messageDigest)` before throwing

Reply:
Done